m2vmp2cut.1
.TH m2vmp2cut 1 2012-02-22
.SH NAME
m2vmp2cut \- MPEG2 frame accurate cutter
.SH SYNOPSIS
.B m2vmp2cut
.I file/directory
.I command
.RI [ argument "... ]"
.SH DESCRIPTION
.B M2vmp2cut
is frame accurate (currently PAL) mpeg2 video (m2v file) with
accompanied mp2 audio (mp2 file) cutter.
Frame accuracy is achieved by re-encoding video around cutpoints.
Audio is cut from separate mp2 file at positions that keep a/v sync as
good as possible (maximum sync difference is around 10-15 milliseconds).

The file/directory argument points which content is to be processed.
\fBdemux\fP needs file argument. All other commands can take either --
directory is resolved from file name if file argument is given.

.TP
\fBdemux\fP [\fIProjectX option\fP...]
.B M2vmp2cut
requires mpeg files to be demuxed to elementary streams before
cutting.
This command uses
.BR ProjectX (1)
to do the demuxing.
A separate directory (based on source filename) is created for output
files.
Note that this doubles the disk usage of a particular source file.
(Not yet counting the space needed for the final output file.)
\fIProjectX option\fPs may contain e.g. more filenames to be demuxed
into output streams.
.TP
\fBselect\fP
This command uses new
.B m2vcut-gui
graphical utility for searching cutpoints.
This work is done frame-accurately.
.TP
\fBcut\fP [\fIoption\fP...]
This command is wrapper to m2vmp2cut.pl (which used to be the frontend
of m2vmp2cut in old versions).
This command has extensive help of it's own.
Note that this (again) adds one third of disk usage (at maximum) so far
when this creates final output file.
.TP
\fBplay\fP [\fIoption\fP...]
This command runs mplayer for the file created with cut command.
.TP
\fBmove\fP \fIdestfile\fP
Moves final output file to a new destination.
.TP
\fBgetyuv\fP [\fBexamples\fP]
Decodes selected mpeg2 frames as a stream of yuv4mpeg pictures.
Useful for further encoding.
If \fBexamples\fP argument is given this command provides example output.

.TP
\fBgetmp2\fP [\fBexamples\fP]
Extracts selected mp2 audio data, to be muxed with (re-encoded) video.
Mp2 data can be used as is, or as encoded to mp3 or vorbis.
If \fBexamples\fP argument is given this command provides example output.
.TP
.B help
Display help message.
.TP
.B x
Affitional tools.
Encoding scripts etc.


NEWS
m2vmp2cut 0.87-wip (unreleased) (Cutting Iron)
==============================================

(2015-02-19)

* m2vcut-gui: When selecting "end" cutpoint of a part the selection is
   inclusive. This makes selecting I or P frame for that more explicit.

* assel-gui: New program to choose audio and subtitle streams; run before
   m2vcut-gui when 'select' command is used.

* contrib has been renamed to pp; the most advanced tools (icut & imkvcut)
   are there. E.g. pp imkvcut can create .mkv file containing one video
   track accompanied with multible audio and subtitle (of pictures) tracks.

* m2vcut-gui: Changed pgup/pgdown keybindings to do 2 min jumps, but after
   first direction change, halve jump in each step.

* m2vcut-gui: Added 6 second jumps to shift+left/right arrow and 1 minute
   jumps to shift+up/down arrow keys.

* m2vcut-gui: The space showing audio levels of current frame now has either
   greenish(frame in) or reddish(frame out) background.

* pp/icut: Support for cutting without re-encoding anything added to
   various places in source code. Sections starting in I-frame and ending
   in I/P-frame can be cut without re-encoding.

* pp/imkvcut: Like icut above, additionally subtitle tracks can be "merged"
   to final file. Also language codes (set by assel-gui) are effective here.

* Some general code improvements and minor behaviour changes done.
* Fine-tuned some build system details.

* ... and probably many things not known until git logs and diff to 0.86
  examimed...

m2vmp2cut 0.86 (2012-10-18)
===========================

* Improved build system with user-definable optimization flags
  and some installation variable by importing and fitting patches
  from Göran Uddeborg.

m2vmp2cut 0.84 (2012-10-14)
===========================

* Added contrib/ffgetyuv.pl which uses ffmpeg to decode m2v input.
* Contrib scripts use ffgetyuv.pl for some m2v error concealment.
* Hid deprecated m2vmp2cut lvev6frames feature.
* Removed obsoleted 'a' (show agraph) feature from select GUI.
* Some small usage/doc improvements.

m2vmp2cut 0.82 (2012-09-30)
===========================

* m2vmp2cut cli now requires source file/dir as first arg (consistently).
* Approximate audio levels visualized in m2vcut.
* Added contrib/webm.sh script.
* Demux output files now written to subdirectory in current dir.
* General code improvemenet with no functional change.

m2vmp2cut 0.78 (2009-11-20) (Lost and Found)
============================================

* If 'projectx' command is found (in PATH), that is used.
* Demux can demux multiple files.
* Bug fix when last chars in file were \000\000\001.

m2vmp2cut 0.76 (2008-09-09)
===========================

* Fixed seeking to offset zero (-sb 0) in m2vtoyuv command.

m2vmp2cut 0.74 (2008-08-18)
===========================

* Fixed largefile support in fileparts command.
* Added getyuv.pl and getmp2.sh -- and related commands to m2vmp2cut;
  these outputs selected video and audio parts from the source.
* Added "contrib" interface.
* Added first "contrib" script, mpeg2 re-encoding.
* Tuned file/directory argument position on m2vmp2cut command (again).

m2vmp2cut 0.72 (2008-07-30)
===========================

* ANNOUNCE informs this is stable version.
* buildlibmpeg-051.sh accepts libmpeg-0.5.1.tar.gz with absolute path.
* Minor tunes, simplification etc.

m2vmp2cut 0.68 (2008-07-27) (Not even far!)
===========================================

* Binaries wavgraph and warpxpointer are installed.
* Audio sync fixed in cutpoint test functionality.
* ProjectX is linked for use during 'make install'.

0.67 (Success is not an option)
'''''''''''''''''''''''''''''''
* Audio visualizer around current position in m2vcut-gui, to find
  (silent) spot for cutpoint.
* Test speedups, now audio is pre-scanned too.
* Some usage tunes, fixes and enchangements.
* Bug fixes in almost every area.

0.65 (Faux pas)
'''''''''''''''
* Target size estimation.
* Keyboard control adjustments in gui.
* Bug fixes.

0.63 (Epic Fail)
''''''''''''''''
* Added graphical gui client for cutpoint selection.
* Added m2vtoyuv for frame decoding -- in place of transcode(1)
* Now gops are closed just after encoded frames.
* Added (cli) frontend to do run demuxer, cutter and m2vmp2cut.pl
*  Lots of other tunes

m2vmp2cut 0.62 (2008-07-27) (Gender-neutral pronouns)
=====================================================

* Makefile tail commands fixed to have -n before +<num>

m2vmp2cut 0.58 (2005-10-23) (Wrecking Stars)
============================================

* Timecodes in GOP's "fixed".
* Audio scanning implemented in C for speed...
  ... and if audio bit/samplerates change in output that will be an
  error.
* Source video format and intra_dc_precision value are used in
  video encoder (Still some settings that cannot be defined in
  encoder left, though).
* Some (a little less significant) improvements

m2vmp2cut 0.56 (2005-03-04) (Fifty-Sixty)
=========================================

* New option: --aonly (cut audio only (video to be re-encoded or so)).
* New option: --evbr (other way to prepare video for re-encoding...).
* Added "encoded frame count" checks.
* Uses lvemedia.lst in addition to lveedit.lst for finding cutpoints
  (to go around (IMHO) lve bug when dragging frame 0).
* Examples tuning/fixing.
* Added some lve patches to ./misc
* Internal workings changes.
* Bug fixes.

m2vmp2cut 0.54 (2004-11-03) (Constit on monet)
==============================================

* --sync option bugfix: negative numbers were taken as positive.
* --test option bugfix: was no-op in 0.42
* New options, --last, --requant, -. And --options, --examples.
* More embedded documentation.
* Added `misc' dir containing external stuff.
* Internal workings changes.

m2vmp2cut 0.42 (2004-10-21) (Infinite Improbabilities)
======================================================

* First release.


unused/m2vmp2cut.c
/*
 * m2vmp2cut.c
 *
 * Author: Tomi Ollila too ät iki fi
 *
 *	Copyright (c) 2004 Tomi Ollila
 *	    All rights reserved
 *
 * Created: Sat Sep 25 09:11:08 EEST 2004 too
 * Last modified: Sun Oct 23 09:17:45 EEST 2005 too
 *
 * This program is licensed under the GPL v2. See file COPYING for details.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DBGS 0
#include "x.h"

#include "ghdrs/m2vmp2cut_priv.h"

/* includes for builtin modules */
#include "m2vscan.h"
#include "mpgfilter.h"
#include "fileparts.h"
#include "mp2cutpoints.h"
#include "filerotate.h"
#include "showdoc.h"
#if ! MINI
#include "embeds.h"
#include "ghdrs/version.h"
#endif


void main(int argc, char ** argv) GCCATTR_NORETURN;
void main(int argc, char ** argv)
{
  if (argc >= 2 && strcmp(argv[1], "builtin") == 0)
    s__builtin(argc, argv);

#if MINI
  xerrf("%s does only `builtin' commands.\n", argv[0]);
#else
  WriteCS(1, "\n*** "); WriteCS(1, version);
  WriteCS(1, "\n"
	  "*** This program is licensed under the GNU Public License v2.\n");

  if (argc < 2)
    {
      WriteCS(2, "\nEnter `");
      write(2, argv[0], strlen(argv[0]));
      WriteCS(2, " --help' for more information.\n\n");
      exit(1);
    }

  WriteCS(1, "*** ...running embedded m2vmp2cut perl driver program...\n");
  run_m2vmp2cut(argc, argv);
#endif
}


static void s__builtin(int argc, char ** argv) /* protoadd GCCATTR_NORETURN */
{
  if (argc >= 3)
    {
#if ! MINI
      if (strcmp(argv[2], "mpg_somehdrinfo.py") == 0)
	builtin_mpg_somehdrinfo(argc, argv);

      if (strcmp(argv[2], "mpg_catfiltered.py") == 0)
	builtin_mpg_catfiltered(argc, argv);

      if (strcmp(argv[2], "lvev6frames.pl") == 0)
	builtin_lvev6frames(argc, argv);
#endif
      if (strcmp(argv[2], "m2vscan") == 0)
	builtin_m2vscan(argc, argv);

      if (strcmp(argv[2], "mpgfilter") == 0)
	builtin_mpgfilter(argc, argv);

      if (strcmp(argv[2], "fileparts") == 0)
	builtin_fileparts(argc, argv);

      if (strcmp(argv[2], "mp2cutpoints") == 0)
	builtin_mp2cutpoints(argc, argv);

      if (strcmp(argv[2], "filerotate") == 0)
	builtin_filerotate(argc, argv);

      if (strcmp(argv[2], "vermatch") == 0)
	s__builtin_vermatch(argc, argv);

      if (strcmp(argv[2], "showdoc") == 0)
	builtin_showdoc(argc, argv);
    }

  xerrf("%s: Unknown builtin command `%s'.\n", argv[0], argv[2]);
}

static void s__builtin_vermatch(int argc, char ** argv)
{
  if (argc == 4 && strcmp(argv[3], "3") == 0)
    exit(0);

  WriteCS(2, "\n!!! m2vmp2cut internal version mismatch !!!\n\n");
  exit(1);
}

/*
 * Local variables:
 * mode: c
 * c-file-style: "gnu"
 * tab-width: 8
 * End:
 */
